### PR TITLE
Core[major]: Base Tracer to propagate raw output from tool for on_tool_end

### PR DESCRIPTION
**Pull request by keenborder786**
This PR completes work for PR #18798 to expose raw tool output in on_tool_end.

Affected APIs:
* astream_log
* astream_events
* callbacks sent to langsmith via langsmith-sdk
* Any other code that relies on BaseTracer!

### CODE REVIEW

**Review by eyurtsev — APPROVED** [review]

**Comment by eyurtsev** [comment]
Waiting for a second maintainer to sign off

**Review by hinthornw — APPROVED** [review]

File: libs/core/tests/unit_tests/runnables/test_runnable_events.py
Comment (eyurtsev) [inline libs/core/tests/unit_tests/runnables/test_runnable_events.py:56]:
```suggestion
    def foo(x: int) -> dict:
```

File: libs/core/tests/unit_tests/runnables/test_runnable_events.py
Comment (eyurtsev) [inline libs/core/tests/unit_tests/runnables/test_runnable_events.py:61]:
```suggestion
    def get_docs(x: int) -> List[Document]:
```

**Review by eyurtsev — APPROVED** [review]

**Comment by eyurtsev** [comment]
Made a change to fix linter errors, ready to merge if tests pass

**Comment by keenborder786** [comment]
@eyurtsev I don't get it why are some checks still failing.

**Comment by PhMueller** [comment]
@eyurtsev 

Hey, 
thanks a lot for your work! 
I just stumbled across this fix here and I am super thankful that you pushed a solution for it.

I'd like to ask if you are planning to apply this change also for the `AsyncCallbackManagerForToolRun`. 

https://github.com/langchain-ai/langchain/blob/d6d843ec24f5ccd29e1f6a4e75fa7ae5ca4a8159/libs/core/langchain_core/callbacks/manager.py#L1041-L1047

**Comment by adreo00** [comment]
@PhMueller This looks like the root cause of issue as well. When I remove this line (see https://github.com/langchain-ai/langchain/pull/20374), the[ code I use to reproduce the issue](https://github.com/langchain-ai/langchain/issues/20372#issue-2239269029) passes.

**Pull request merged**
